[Swift in WebKit] Fix Swift 6 warnings and errors in WKSLinearMediaPlayer

rr-codes opened the pull request:
#### facbe72f14ef35005af22fe8d56ded0105305d8f
<pre>
[Swift in WebKit] Fix Swift 6 warnings and errors in WKSLinearMediaPlayer
<a href="https://bugs.webkit.org/show_bug.cgi?id=294823">https://bugs.webkit.org/show_bug.cgi?id=294823</a>
<a href="https://rdar.apple.com/154057962">rdar://154057962</a>

Reviewed by Andy Estes.

Address Swift 6 errors in WKSLinearMediaPlayer and related code.

- If it walks like a duck and talks like a duck... it&apos;s not always a duck. In this case, AVKit has a public
Swift class `PlayableViewController` and annotated with `@objc(LMPlayableViewController)`. This annotation
makes the type compatible and indistuinguishable from an Objective C type. However, Swift types cannot be accessed
by Objective-C in another framework; within their own framework, the Swift bridging header is used.

AVKit attempted to work around this by creating an Objective-C interface that was essentially a forward declaration
of `LMPlayableViewController`; this is an unsupported configuration as it removes all compile-time checking, which
results in the Swift compiler being confused.

To fix, the `LMPlayableViewController` type is now never used within WebKit; only the `PlayableViewController` type
is used from within Swift WebKit code. Consequently, a new wrapper type is created that abstracts the
`PlayableViewController` and exposes the needed symbols to Objective-C, so that the compiler can function properly.

- Use `@objc @implementation` instead of `@_objcImplementation`.

- Use `RealityKit` instead of the re-exported `RealityFoundation` module.

- Adopt `@preconcurrency` when needed, to silence errors due to the underlying framework symbols
not having proper annotations.

* Source/WebCore/platform/ios/VideoPresentationInterfaceAVKit.h:
* Source/WebCore/platform/ios/VideoPresentationInterfaceIOS.h:
(WebCore::VideoPresentationInterfaceIOS::playableViewController):
* Source/WebKit/Platform/ios/VideoPresentationInterfaceLMK.h:
* Source/WebKit/Platform/ios/VideoPresentationInterfaceLMK.mm:
(WebKit::VideoPresentationInterfaceLMK::setupFullscreen):
(WebKit::VideoPresentationInterfaceLMK::enterExternalPlayback):
(WebKit::VideoPresentationInterfaceLMK::playerViewController const):
(WebKit::VideoPresentationInterfaceLMK::playableViewController):
(WebKit::VideoPresentationInterfaceLMK::ensurePlayableViewController):
* Source/WebKit/UIProcess/Cocoa/VideoPresentationManagerProxy.h:
* Source/WebKit/UIProcess/Cocoa/VideoPresentationManagerProxy.mm:
(WebKit::VideoPresentationManagerProxy::playableViewController const):
* Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenViewController.mm:
(-[WKFullScreenViewController configureEnvironmentPickerOrFullscreenVideoButtonView]):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebKitSwift/LinearMediaKit/LinearMediaKitExtras.h: Removed.
* Source/WebKit/WebKitSwift/LinearMediaKit/LinearMediaKitExtras.swift: Removed.
* Source/WebKit/WebKitSwift/LinearMediaKit/LinearMediaKitSPI.h: Removed.
* Source/WebKit/WebKitSwift/LinearMediaKit/LinearMediaPlayer.swift:
(SwiftOnlyData.viewController):
(WKSLinearMediaPlayer.makeViewController):
(WKSLinearMediaPlayer.enterExternalPresentation(_:(any Error)?) -&gt; Void:)):
(WKSLinearMediaPlayer.exitExternalPresentation(_:(any Error)?) -&gt; Void:)):
(WKSLinearMediaPlayer.enterFullscreen(_:(any Error)?) -&gt; Void:)):
(WKSLinearMediaPlayer.exitFullscreen(_:(any Error)?) -&gt; Void:)):
(WKSLinearMediaPlayer.maybeCreateSpatialOrImmersiveEntity):
(WKSLinearMediaPlayer.maybeClearSpatialOrImmersiveEntity):
* Source/WebKit/WebKitSwift/LinearMediaKit/LinearMediaTypes.swift:
(WKSPlayableViewControllerHost.baseViewController):
(WKSPlayableViewControllerHost.automaticallyDockOnFullScreenPresentation):
(WKSPlayableViewControllerHost.dismissFullScreenOnExitingDocking):
(WKSPlayableViewControllerHost.environmentPickerButtonViewController):
(WKSPlayableViewControllerHost.playable):
(WKSPlayableViewControllerHost.prefersAutoDimming):
* Source/WebKit/WebKitSwift/LinearMediaKit/WKSLinearMediaPlayer.h:
* Source/WebKit/WebKitSwift/LinearMediaKit/WKSLinearMediaTypes.h:
* Source/WebKit/WebKitSwift/WebKitSwift.h:

Canonical link: <a href="https://commits.webkit.org/296600@main">https://commits.webkit.org/296600@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/33e59b56d7384fff4ff74a83fb3e3cdf718eeac0

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/109007 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/28668 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/19092 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/114218 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/59330 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/110970 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/29350 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/37234 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/82840 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/111955 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/23345 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/98186 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/63282 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/22754 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/16325 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/58910 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/92713 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/16371 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/117336 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/36056 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/26651 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/91855 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/36426 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/94450 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/91660 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23336 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/36569 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/14320 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/31914 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/35953 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/41468 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/35649 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/38989 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/37336 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aestes commented on Source/WebCore/platform/ios/VideoPresentationInterfaceIOS.h:
```diff
@@ -47,7 +47,7 @@
 
 OBJC_CLASS AVPlayerViewController;
-OBJC_CLASS LMPlayableViewController;
+OBJC_CLASS WKPlayableViewController;
```
Nit: mis-sorted

aestes commented on Source/WebKit/Platform/ios/VideoPresentationInterfaceLMK.h:
```diff
@@ -30,7 +30,7 @@
 #include <wtf/TZoneMalloc.h>
 
-OBJC_CLASS LMPlayableViewController;
+OBJC_CLASS WKPlayableViewController;
```
Nit: mis-sorted

aestes commented on Source/WebKit/Platform/ios/VideoPresentationInterfaceLMK.mm:
```diff
@@ -300,7 +298,7 @@ - (void)layoutSublayers
     ALWAYS_LOG_IF_POSSIBLE(LOGIDENTIFIER);
     m_playerViewController = [linearMediaPlayer() makeViewController];
-    [m_playerViewController view].alpha = 0;
+    [[m_playerViewController baseViewController] view].alpha = 0;
```
Nit:
```suggestion
    [m_playerViewController baseViewController].view.alpha = 0;
```

aestes commented on Source/WebKit/UIProcess/Cocoa/VideoPresentationManagerProxy.h:
```diff
@@ -45,7 +45,7 @@
 #include <wtf/text/WTFString.h>
 
-OBJC_CLASS LMPlayableViewController;
+OBJC_CLASS WKPlayableViewController;
```
Nit: mis-sorted

aestes commented on Source/WebKit/WebKitSwift/LinearMediaKit/WKSLinearMediaTypes.h:
```diff
@@ -98,6 +94,19 @@ API_AVAILABLE(visionos(1.0))
 
-NS_ASSUME_NONNULL_END
+NS_SWIFT_UI_ACTOR
+@interface WKPlayableViewController : NSObject
```
I think this should be called `WKSPlayableViewController` so long as it lives in WebKitSwift. It's also strange to call this a view controller when it isn't (i.e. is not a `UIViewController`).

aestes commented on Source/WebCore/platform/ios/VideoPresentationInterfaceIOS.h:
```diff
@@ -47,14 +47,14 @@
 OBJC_CLASS CALayer;
 OBJC_CLASS NSError;
 OBJC_CLASS WebAVPlayerController;
+OBJC_CLASS WKSPlayableViewControllerHost;
```
Nit: mis-sorted

aestes commented on Source/WebKit/Platform/ios/VideoPresentationInterfaceLMK.h:
```diff
@@ -81,14 +81,14 @@ class VideoPresentationInterfaceLMK final : public WebCore::VideoPresentationInt
     void ensurePlayableViewController();
 
-    RetainPtr<LMPlayableViewController> m_playerViewController;
+    RetainPtr<WKSPlayableViewControllerHost> m_playerViewController;
```
```suggestion
    RetainPtr<WKSPlayableViewControllerHost> m_playerViewControllerHost;
```

aestes commented on Source/WebKit/UIProcess/Cocoa/VideoPresentationManagerProxy.h:
```diff
@@ -186,7 +186,7 @@ class VideoPresentationManagerProxy
 
 #if ENABLE(LINEAR_MEDIA_PLAYER)
-    LMPlayableViewController *playableViewController(PlaybackSessionContextIdentifier) const;
+        WKSPlayableViewControllerHost *playableViewController(PlaybackSessionContextIdentifier) const;
```
```suggestion
    WKSPlayableViewControllerHost *playableViewController(PlaybackSessionContextIdentifier) const;
```

aestes commented on Source/WebKit/WebKitSwift/LinearMediaKit/WKSLinearMediaTypes.h:
```diff
@@ -98,6 +94,19 @@ API_AVAILABLE(visionos(1.0))
+NS_SWIFT_UI_ACTOR
+@interface WKSPlayableViewControllerHost : NSObject
+
+@property (nonatomic, readonly) UIViewController *baseViewController;
```
Would be OK to just call this `viewController` or `playableViewController` now that the class has a better name.

rr-codes committed:
[Swift in WebKit] Fix Swift 6 warnings and errors in WKSLinearMediaPlayer
https://bugs.webkit.org/show_bug.cgi?id=294823
rdar://154057962

Reviewed by Andy Estes.

Address Swift 6 errors in WKSLinearMediaPlayer and related code.

- If it walks like a duck and talks like a duck... it's not always a duck. In this case, AVKit has a public
Swift class `PlayableViewController` and annotated with `@objc(LMPlayableViewController)`. This annotation
makes the type compatible and indistuinguishable from an Objective C type. However, Swift types cannot be accessed
by Objective-C in another framework; within their own framework, the Swift bridging header is used.

AVKit attempted to work around this by creating an Objective-C interface that was essentially a forward declaration
of `LMPlayableViewController`; this is an unsupported configuration as it removes all compile-time checking, which
results in the Swift compiler being confused.

To fix, the `LMPlayableViewController` type is now never used within WebKit; only the `PlayableViewController` type
is used from within Swift WebKit code. Consequently, a new wrapper type is created that abstracts the
`PlayableViewController` and exposes the needed symbols to Objective-C, so that the compiler can function properly.

- Use `@objc @implementation` instead of `@_objcImplementation`.

- Use `RealityKit` instead of the re-exported `RealityFoundation` module.

- Adopt `@preconcurrency` when needed, to silence errors due to the underlying framework symbols
not having proper annotations.

* Source/WebCore/platform/ios/VideoPresentationInterfaceAVKit.h:
* Source/WebCore/platform/ios/VideoPresentationInterfaceIOS.h:
(WebCore::VideoPresentationInterfaceIOS::playableViewController):
* Source/WebKit/Platform/ios/VideoPresentationInterfaceLMK.h:
* Source/WebKit/Platform/ios/VideoPresentationInterfaceLMK.mm:
(WebKit::VideoPresentationInterfaceLMK::setupFullscreen):
(WebKit::VideoPresentationInterfaceLMK::enterExternalPlayback):
(WebKit::VideoPresentationInterfaceLMK::playerViewController const):
(WebKit::VideoPresentationInterfaceLMK::playableViewController):
(WebKit::VideoPresentationInterfaceLMK::ensurePlayableViewController):
* Source/WebKit/UIProcess/Cocoa/VideoPresentationManagerProxy.h:
* Source/WebKit/UIProcess/Cocoa/VideoPresentationManagerProxy.mm:
(WebKit::VideoPresentationManagerProxy::playableViewController const):
* Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenViewController.mm:
(-[WKFullScreenViewController configureEnvironmentPickerOrFullscreenVideoButtonView]):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebKitSwift/LinearMediaKit/LinearMediaKitExtras.h: Removed.
* Source/WebKit/WebKitSwift/LinearMediaKit/LinearMediaKitExtras.swift: Removed.
* Source/WebKit/WebKitSwift/LinearMediaKit/LinearMediaKitSPI.h: Removed.
* Source/WebKit/WebKitSwift/LinearMediaKit/LinearMediaPlayer.swift:
(SwiftOnlyData.viewController):
(WKSLinearMediaPlayer.makeViewController):
(WKSLinearMediaPlayer.enterExternalPresentation(_:(any Error)?) -> Void:)):
(WKSLinearMediaPlayer.exitExternalPresentation(_:(any Error)?) -> Void:)):
(WKSLinearMediaPlayer.enterFullscreen(_:(any Error)?) -> Void:)):
(WKSLinearMediaPlayer.exitFullscreen(_:(any Error)?) -> Void:)):
(WKSLinearMediaPlayer.maybeCreateSpatialOrImmersiveEntity):
(WKSLinearMediaPlayer.maybeClearSpatialOrImmersiveEntity):
* Source/WebKit/WebKitSwift/LinearMediaKit/LinearMediaTypes.swift:
(WKSPlayableViewControllerHost.baseViewController):
(WKSPlayableViewControllerHost.automaticallyDockOnFullScreenPresentation):
(WKSPlayableViewControllerHost.dismissFullScreenOnExitingDocking):
(WKSPlayableViewControllerHost.environmentPickerButtonViewController):
(WKSPlayableViewControllerHost.playable):
(WKSPlayableViewControllerHost.prefersAutoDimming):
* Source/WebKit/WebKitSwift/LinearMediaKit/WKSLinearMediaPlayer.h:
* Source/WebKit/WebKitSwift/LinearMediaKit/WKSLinearMediaTypes.h:
* Source/WebKit/WebKitSwift/WebKitSwift.h:

Canonical link: https://commits.webkit.org/296600@main
